go/interpreter/lfvm/instructions_test.go

```diff
@@ -2036,6 +1912,86 @@ func TestOpExtCodeHash_WritesHashOnStackIfAccountExists(t *testing.T) {
 	}
 }
 
+func TestInstructions_MCopy_ReturnsErrorOnFailure(t *testing.T) {
+
+	tests := map[string]struct {
+		destOffset, offset, size uint64
+		expectedError            error
+		gasRemoved               uint64
+	}{
+		"does nothing if size is zero": {
+			destOffset: 1024, offset: 125, size: 0,
+		},
+		"returns error when failed read memory expansion": {
+			destOffset: 0, offset: math.MaxUint64, size: 1,
+			expectedError: errOverflow,
+		},
+		"returns error when failed write memory expansion": {
+			destOffset: math.MaxUint64, offset: 0, size: 1,
+			expectedError: errOverflow,
+		},
+		"returns error if gas is not enough for size": {
+			destOffset: 0, offset: 0, size: 128,
+			expectedError: errOutOfGas,
+			gasRemoved:    1,
+		},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			ctxt := getEmptyContext()
+			ctxt.params.Revision = tosca.R13_Cancun
+			ctxt.stack = fillStack(
+				*uint256.NewInt(test.destOffset),
+				*uint256.NewInt(test.offset),
+				*uint256.NewInt(test.size),
+			)
+
+			// ignore memory setup errors, to focus on the mcopy operation
+			// expansion is done to accumulate memory cost and focus on the
+			// word count gas cost.
+			_ = ctxt.memory.expandMemory(test.destOffset, test.size, &context{gas: 1 << 32})
+			_ = ctxt.memory.expandMemory(test.offset, test.size, &context{gas: 1 << 32})
+			ctxt.gas = tosca.Gas(3*tosca.SizeInWords(test.size) - test.gasRemoved)
+
+			err := opMcopy(&ctxt)
+			if err != test.expectedError {
+				t.Fatalf("unexpected error, wanted %v, got %v", test.expectedError, err)
+			}
+		})
+	}
+}
+
+func TestInstructions_MCopy_CopiesOverlappingRanges(t *testing.T) {
+
+	ctxt := getEmptyContext()
+	ctxt.params.Revision = tosca.R13_Cancun
+	ctxt.stack = fillStack(
+		*uint256.NewInt(5),
+		*uint256.NewInt(0),
+		*uint256.NewInt(10))
+
+	err := ctxt.memory.set(
+		uint256.NewInt(0),
+		[]byte{0, 1, 2, 3, 4, 5, 6, 7, 8, 9},
+		&context{gas: 1 << 32},
+	)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	ctxt.gas = tosca.Gas(3 * tosca.SizeInWords(10))
+
+	err = opMcopy(&ctxt)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	expected := []byte{0, 1, 2, 3, 4, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9}
+	if want, got := expected, ctxt.memory.store[0:15]; !bytes.Equal(want, got) {
```

Comment:
why not check the whole storage slice ?

Reply:
the memory copied is the one in the range defined. This is needed to keep the test generic, if we add a new test case this code must still be capable of executing both cases.